libsolidity/ast/AST.cpp

```diff
@@ -374,6 +374,17 @@ std::vector<std::pair<ASTPointer<IdentifierPath>, std::optional<Token>>> UsingFo
 	return ranges::zip_view(m_functionsOrLibrary, m_operators) | ranges::to<vector>;
 }
 
+util::h256 StructDefinition::typehash() const
+{
+	std::string str = name() + "(";
+	for (size_t i = 0; i < m_members.size(); i++)
+	{
+		str += i == 0 ? "" : ",";
+		str += m_members[i]->type()->canonicalName() + " " + m_members[i]->name();
+	}
+	return util::keccak256(str + ")");
+}
```

Comment:
How about recursive types?

I think there are also types that cannot have a typehash, for example if it contains a member that cannot be ABI-encoded such as internal function pointers / mapping types.

We'll need to handle those cases in a typechecker.

Reply:
Also the inner struct should be sorted by name and appended.

Reply:
I see nested structures should be described as following (https://eips.ethereum.org/EIPS/eip-712):
> If the struct type references other struct types (and these in turn reference even more struct types), then the set of referenced struct types is collected, sorted by name and appended to the encoding. An example encoding is `Transaction(Person from,Person to,Asset tx)Asset(address token,uint256 amount)Person(address wallet,string name)`.

Reply:
Added commit to solve this behavior, still not sure about this case: `A(B b,C c)B(D d)D(uint256)C(D d)D(uint256)`, type `D` will be mentioned twice. Not sure is it issue or not and not sure what EIP-712 should do about this.

Reply:
Yes, `EIP-712` is a bit vague there, but I think that should be transformed into:

```solidity
"A(B b,C c)B(D d)C(D d)D(uint256 x)"
```

Reply:
Agree, that’s what I tried to implement in the latest commit.